pkg/apis/core/types.go

```diff
@@ -5763,3 +5771,444 @@ type PortStatus struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Items is the list of resource claims.
+	Items []PodScheduling `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ResourceClass is used by administrators to influence how resources
+// are allocated.
+type ResourceClass struct {
+	metav1.TypeMeta
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta
+
+	// DriverName determines which resource driver is to be used for
+	// allocation of a ResourceClaim that uses this class.
+	//
+	// Resource drivers have a unique name in reverse domain order
+	// (acme.example.com).
+	DriverName string
+
+	// Parameters holds arbitrary values that will be available to the
+	// driver when allocating a resource that uses this class. The driver
+	// will be able to distinguish between parameters stored here and and
+	// those stored in ResourceClaimSpec. These parameters here can only be
+	// set by cluster administrators.
+	Parameters *ResourceClassParametersReference
+
+	// Only nodes matching the selector will be considered by the scheduler
+	// when trying to find a Node that fits a Pod when that Pod uses
+	// a ResourceClaim that has not been allocated yet.
+	//
+	// Setting this field is optional. If nil, all nodes are candidates.
+	SuitableNodes *NodeSelector
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ResourceClassList is a collection of resource classes.
+type ResourceClassList struct {
+	metav1.TypeMeta
+	// Standard list metadata
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ListMeta
+
+	// Items is the list of resource classes.
+	Items []ResourceClass
+}
+
+// ResourceClassParametersReference contains enough information to let you
+// locate the parameters for a ResourceClass. The object must be
+// cluster-scoped.
+type ResourceClassParametersReference struct {
+	// APIVersion is the group and version for the resource being referenced or
+	// just the version for the core API.
+	APIVersion string
+	// Kind is the type of resource being referenced. This is the same
+	// value as in the parameter object's metadata.
+	Kind string
+	// Name is the name of resource being referenced.
+	Name string
+}
+
+// ResourceClaimParametersReference contains enough information to let you
+// locate the parameters for a ResourceClaim. The object must be in the same
+// namespace as the ResourceClaim.
+type ResourceClaimParametersReference struct {
+	// APIVersion is the group and version for the resource being referenced or
+	// just the version for the core API.
+	APIVersion string
```

Comment:
I switched to `APIVersion` (e.g. 'v1' for `ConfigMap`) here instead of `APIGroup` for the sake of consistency with what users put into their YAML files when defining the parameters.

But I also see that this slightly overspecifies the reference here because the resource driver can decide to read as a different version than the one created by the user. From that perspective, `APIGroup` makes more sense.

When using `APIGroup`, I would make it a plain string, not a string pointer, and treat it as optional. This is different from APIGroup in TypedLocalObjectReference, see https://kubernetes.slack.com/archives/CJUQN3E4T/p1657203451006569

Reply:
I switched back to `APIGroup`. It's more correct from a technical perspective and consistent.